src/core/server/saved_objects/migrations/core/migration_context.ts

```diff
@@ -107,20 +111,68 @@ function createSourceContext(source: FullIndexInfo, alias: string) {
 function createDestContext(
   source: FullIndexInfo,
   alias: string,
-  mappingProperties: SavedObjectsTypeMappingDefinitions
+  typeMappingDefinitions: SavedObjectsTypeMappingDefinitions
 ): FullIndexInfo {
-  const activeMappings = buildActiveMappings(mappingProperties);
+  const targetMappings = disableUnknownTypeMappingFields(
+    buildActiveMappings(typeMappingDefinitions),
+    source.mappings
+  );
 
   return {
     aliases: {},
     exists: false,
     indexName: nextIndexName(source.indexName, alias),
-    mappings: {
-      ...activeMappings,
-      properties: {
-        ...source.mappings.properties,
-        ...activeMappings.properties,
-      },
+    mappings: targetMappings,
+  };
+}
+
+/**
+ * Merges the active mappings and the source mappings while disabling the
+ * fields of any unknown Saved Object types present in the source index's
+ * mappings.
+ *
+ * Since the Saved Objects index has `dynamic: strict` defined at the
+ * top-level, only Saved Object types for which a mapping exists can be
+ * inserted into the index. To ensure that we can continue to store Saved
+ * Object documents belonging to a disabled plugin we define a mapping for all
+ * the unknown Saved Object types that were present in the source index's
+ * mappings. To limit the field count as much as possible, these unkwnown
+ * type's mappings are set to `dynamic: false`.
+ *
+ * (Since we're using the source index mappings instead of looking at actual
+ * document types in the inedx, we potentially add more "unknown types" than
+ * what would be necessary to support migrating all the data over to the
+ * target index.)
+ *
+ * @param activeMappings The mappings compiled from all the Saved Object types
+ * known to this Kibana node.
+ * @param sourceMappings The mappings of index used as the migration source.
+ * @returns The mappings that should be applied to the target index.
+ */
+function disableUnknownTypeMappingFields(
+  activeMappings: IndexMapping,
+  sourceMappings: IndexMapping
+): IndexMapping {
+  const targetTypes = Object.keys(activeMappings.properties);
```

Comment:
Implem looks alright to me, but I would add unit test on this specific function in addition to the additions to `index_migrator.test.ts` as it's quite a sensible piece of code.